### PR TITLE
New generate_query_graph tool.

### PR DESCRIPTION
**Pull request by Tmonster**
The generate query graph code was taken out in https://github.com/duckdb/duckdb/pull/6972. Here I'm attempting to add it  back in the generate query graph code. Using some pastel colors but it can honestly be anything. 

Some goals for this version of the query graph 
- [X] *No javascript*. Javascript is hard to maintain. This tree library is basically all CSS and doesn't need maintenance
- [X] *Readability*. The previous query graphs were not very readable (some colors made the text hard to read), The new colors prevent this from happening.
- [X]  Same timing information as the previous version. 

Things to improve
- [ ] The ability to zoom out and see the whole plan. If the plan is bushy, the graph can get very wide, and it becomes difficult to see all of the operators in one screen. Currently however, it is much easier for a user to just navigate across the whole plan. This can also be fixed by decreasing the font for the html.
- [ ] When detailed profiling is provided, add another UI element to show the timing information for the individual optimizers
- [ ] better colors?

open to some initial feedback right now about colors/design. 
You can check out the branch and generate the query graph with the following commands while in the duckdb directory.

```sh
duckdb -c "load tpch; call dbgen(sf=0.1); pragma enable_profiling=json;pragma profile_output="q09.json";pragma tpch(9);"
python3 scripts/generate_querygraph.py q09.json
```

Old query graph generator output 
image

New query graph generator output
image

### CODE REVIEW

File: scripts/generate_querygraph.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        print("Incorrect input for open_output, expected TRUE or FALSE")
+        exit(1)
+
+generate(input, output)
+
+with open(output, 'r') as f:
+    text = f.read()
+
+if open_output:
+    os.system('open "' + output.replace('"', '\\"') + '"')
```
Comment (Mause) [inline scripts/generate_querygraph.py:45]: Let's use the webbrowser module here instead

File: tools/pythonpkg/duckdb/__init__.py [hunk 2]
```diff
@@ -3,9 +3,11 @@
 # Modules
 import duckdb.functional as functional
 import duckdb.typing as typing
+import duckdb.duckdb_query_graph as duckdb_query_graph
 _exported_symbols.extend([
     "typing",
-    "functional"
+    "functional",
+    "duckdb_query_graph"
```
Comment (Mause) [inline tools/pythonpkg/duckdb/__init__.py:10]: I don't think the module name needs to include duckdb

File: tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+		phases.reverse()
+		return phases
+
+	def get_sum_of_all_timings(self):
+		total_timing_sum = 0
+		for phase in self.phase_to_timings.keys():
+			total_timing_sum += self.get_summary_phase_timings(phase).time
+		return total_timing_sum
+
+QUERY_TIMINGS = AllTimings()
```
Comment (Mause) [inline tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py:92]: I don't see this being reset anywhere, can the graph generator only be used once per interpreter session?

File: scripts/generate_querygraph.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# generate_querygraph.py
```
Comment (Mause) [inline scripts/generate_querygraph.py:1]: Let's include this code and ship it as an executable module instead

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py (hunk 3 above)
Comment (Tmonster) [reply tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py:92]: That's the idea yea, but it's probably good to just reset this every time a new profile output is passed

Comment (Mause) [reply tools/pythonpkg/duckdb/duckdb_query_graph/__init__.py:92]: Better to not use a global to begin with probably?

File: scripts/generate_querygraph.py (hunk 4 above)
Comment (Tishj) [reply scripts/generate_querygraph.py:1]: So that means it runs as `python3 -m generate_querygraph explain_output.json` ?

Comment (Mause) [reply scripts/generate_querygraph.py:1]: Like

```sh
python -m duckdb.query_graph ...
```

But basically yeah. Could even publish it an standalone executable on the user machines that way too

File: scripts/generate_querygraph.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import os
+import sys
+from functools import reduce
+from python_helpers import open_utf8
+
+from duckdb.duckdb_query_graph import generate
+
+sys.path.insert(0, 'benchmark')
+
+arguments = sys.argv
```
[the remark this answers is not in the document]
Comment (Mause) [reply scripts/generate_querygraph.py:14]: 💯

**Comment by Tmonster** [comment]
Pressed the re-review button too early. Wait a couple of minutes

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+    open_output = args.open
+
+    translate_json_to_html(input, output)
+
+    with open(output, 'r') as f:
+        text = f.read()
+
+    if open_output:
+        webbrowser.open('file:///Users/tomebergen/scripts/run_imdb/tpch_new_html/q08.html', new=2)
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:319]: Forgot to replace this?

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        else:
+            print("please provide valid .html file for output name")
+            exit(1)
+
+    open_output = args.open
+
+    translate_json_to_html(input, output)
+
+    with open(output, 'r') as f:
+        text = f.read()
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:316]: What's this for?

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    css += qgraph_css + "\n"
+    css += "</style>\n"
+    return {
+        'treeflex_css': treeflex_css,
+        'duckdb_css': css,
+        'libraries': '',
+        'chart_script': ''
+    }
+
+# ????
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:240]: ????

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+    if sys.version_info[0] < 3:
+        return open(fpath, flags)
+    else:
+        return open(fpath, flags, encoding="utf8")
+
+
+# if detailed profiling is enabled, then information on the timing of each stage of the query is available
+# that is "top level timing information"
+def get_top_level_timings(json):
+    return []
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:106]: Is this supposed to contain something?

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+    def get_sum_of_all_timings(self):
+        total_timing_sum = 0
+        for phase in self.phase_to_timings.keys():
+            total_timing_sum += self.get_summary_phase_timings(phase).time
+        return total_timing_sum
+
+
+def open_utf8(fpath, flags):
+    import sys
+    if sys.version_info[0] < 3:
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:97]: We don't support python 2 anymore

File: scripts/generate_querygraph.py (hunk 4 above)
Comment (Mause) [inline scripts/generate_querygraph.py:1]: This file can be deleted now right?

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/duckdb/query_graph/__main__.py (hunk 10 above)
Comment (Mause) [reply tools/pythonpkg/duckdb/query_graph/__main__.py:97]: I think you can just inline this function, no warning required

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+        'duckdb_css': css,
+        'libraries': '',
+        'chart_script': ''
+    }
+
+def gather_timing_information(json, query_timings):
+    # add up all of the times
+    # measure each time as a percentage of the total time.
+    # then you can return a list of [phase, time, percentage]
+    child_timings = get_child_timings(json['children'][0], query_timings)
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:239]: This variable is never used, just call the function

File: tools/pythonpkg/duckdb/query_graph/__main__.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+</html>
+"""
+        html = html.replace("${TREEFLEX_CSS}", html_output['treeflex_css'])
+        html = html.replace("${DUCKDB_CSS}", html_output['duckdb_css'])
+        html = html.replace("${TIMING_TABLE}", timing_table)
+        html = html.replace('${TREE}', tree_output)
+        f.write(html)
+
+def main():
+    sys.path.insert(0, 'benchmark')
```
Comment (Mause) [inline tools/pythonpkg/duckdb/query_graph/__main__.py:279]: Why have you added this line?

**Review by Mause — CHANGES_REQUESTED** [review]

**Comment by Tmonster** [comment]
@Mause Some of those lines are leftover from the old script which I dropped in. Sorry for not thoroughly checking it. Thank you for the reviews!

**Review by Mause — CHANGES_REQUESTED** [review]
Looks good. Just need to run the `make format-fix` script and update for the stubs test

**Comment by Mytherin** [comment]
@Tmonster @Mause is this ready to merge?

**Comment by Tmonster** [comment]
Ready from my side 👍

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**